File: lib/core/external_link.dart
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

import '../design_system/colors.dart';

/// Schemes we allow to leave the app. Imported notes/links can carry arbitrary
/// hrefs (`tel:`, `sms:`, custom app schemes, …); we only ever hand a URL to the
/// OS when it uses one of these safe, expected schemes.
const _allowedSchemes = {'http', 'https', 'mailto'};

/// R4: single, security-conscious entry point for opening a URL that came from
/// importable content. Parses [href], rejects anything outside
/// [_allowedSchemes] (disallowed/unparseable → friendly snackbar, no launch),
/// then launches the rest in the external application.
Future<void> launchExternal(BuildContext context, String href) async {
  final uri = Uri.tryParse(href.trim());
  if (uri == null || !_allowedSchemes.contains(uri.scheme.toLowerCase())) {
    _showBlockedSnackBar(context);
    return;
  }
  final ok = await launchUrl(uri, mode: LaunchMode.externalApplication);
  if (!ok && context.mounted) {
    _showBlockedSnackBar(context);
  }
}

void _showBlockedSnackBar(BuildContext context) {
  if (!context.mounted) return;
  ScaffoldMessenger.of(context).showSnackBar(
    const SnackBar(
      content: Text("Couldn't open that link."),
      backgroundColor: AppColors.accentDanger,
    ),
  );
}

File: webapp/src/core/externalLink.ts
import { showSnackbar } from './snackbar';

const ALLOWED_SCHEMES = new Set(['http:', 'https:', 'mailto:']);

/**
 * Allow-list check for URLs that came from importable content.
 * Returns the normalized href when the scheme is http/https/mailto,
 * or null for anything unparseable or disallowed (javascript:, file:, …).
 */
export function allowedExternalUrl(href: string): string | null {
  let url: URL;
  try {
    url = new URL(href.trim());
  } catch {
    return null;
  }
  if (!ALLOWED_SCHEMES.has(url.protocol.toLowerCase())) return null;
  return url.href;
}

/**
 * Security-conscious single entry point for opening external URLs.
 * Mirrors lib/core/external_link.dart — disallowed/unopenable links surface
 * the exact "Couldn't open that link." snackbar and nothing else happens.
 */
export function launchExternal(href: string): boolean {
  const url = allowedExternalUrl(href);
  if (url === null) {
    showSnackbar("Couldn't open that link.", { danger: true });
    return false;
  }
  if (url.startsWith('mailto:')) {
    // window.open on mailto leaves a blank tab behind in most browsers.
    window.location.href = url;
    return true;
  }
  const opened = window.open(url, '_blank', 'noopener,noreferrer');
  if (opened === null) {
    showSnackbar("Couldn't open that link.", { danger: true });
    return false;
  }
  return true;
}
